torch_glow/src/PyTorchFileLoader.cpp

```diff
@@ -15,30 +15,189 @@
+    torch::jit::RegisterOperators op({torch::jit::Operator(
+        getFusionSymbol(),
+        [](const torch::jit::Node *node) {
+          return [node](torch::jit::Stack &stack) {
+            // Get JIT Graph.
+            auto graph = node->g(at::attr::Subgraph);
+            auto err = loadJitGraphToGlowFunction(
+                stack, *graph, *localFusionInfo.function,
+                *localFusionInfo.inputPlaceholders,
+                *localFusionInfo.outputPlaceholders, *localFusionInfo.settings);
+            if (static_cast<bool>(err)) {
+              throw std::invalid_argument(llvm::toString(std::move(err)));
+            }
+            return 0;
+          };
+        },
+        options)});
+
+    torch::jit::RegisterPass pass([&](std::shared_ptr<torch::jit::Graph> &g) {
+      // Trigger custom fusion pass only if local thread Glow Function is set.
+      if (localFusionInfo.function) {
+        glowCustomFuse(g, getFusionSymbol());
+      }
+    });
+  }
+};
+} // namespace
+
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchModel(
+    const std::string &fileName,
+    std::shared_ptr<torch::jit::script::Module> &module) {
+  try {
+    module = std::make_shared<torch::jit::script::Module>(
+        torch::jit::load(fileName));
+  } catch (const std::exception &x) {
+    RETURN_ERR(strFormat("Cannot load model from file: %s, , reason: %s",
+                         fileName.c_str(), x.what()));
+  }
+  return llvm::Error::success();
+}
+
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchGraph(
+    const std::string &fileName, std::vector<torch::jit::IValue> &inputs,
+    glow::Function &F, std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders,
+    const PyTorchLoaderSettings &settings, bool sanityCheck) {
+  // Register custom pass once.
+  static RegisterCustomFusionPass fuser;
+
+  // Convert PyTorch model into JIT Module.
+  std::shared_ptr<torch::jit::script::Module> module;
+  RETURN_IF_ERR(PyTorchFileLoader::loadPyTorchModel(fileName, module));
+
+  // Disable gradient nodes generation.
+  at::NoGradGuard guard;
+
+  // Set thread local pointer, not null values activate custom fusion pass.
+  localFusionInfo.set(&F, &inputPlaceholders, &outputPlaceholders, &settings);
+  auto err = evaluateModuleGraph(module, inputs);
+  localFusionInfo.reset();
+
+  RETURN_IF_ERR(err);
+
+  return sanityCheck ? performSanityCheck() : llvm::Error::success();
+}
+
+/*static*/
+llvm::Error PyTorchFileLoader::performSanityCheck() {
+  // Sanity check, after "fusionSymbol" node, not other nodes should exist.
+  std::shared_ptr<torch::jit::Graph> subgraph =
+      torch::jit::lastExecutedOptimizedGraph();
+  size_t fusedNodes = 0, missedNodes = 0;
+  const auto symbol = getFusionSymbol();
+  for (const auto &node : subgraph->nodes()) {
+    if (node->kind() == symbol) {
+      ++fusedNodes;
+    } else if (fusedNodes) {
```

Comment:
This will only work if the non-fused nodes are after the fusion group. Is this the intention? if so, why?

Reply:
Normally - ALL operators must be fused into one graph node. If you take a look at graph it contains from the beginning the original graph nodes and then fused node, even all nodes got fused.

Reply:
Oh whoa, that seems super weird. I think I see why that's happening, but why is that desirable instead of deleting the PT graph nodes once they've been fused into Glow?